docs/zkapps/tutorials/05-common-types-and-functions.mdx

```diff
@@ -0,0 +1,360 @@
+
+See the following for an example of how to use this, implementing a `Points` structure, and an array of points of length 8 structure:
+
+```ts
+class Point extends CircuitValue {
+  @prop x: Field;
+  @prop y: Field;
+
+  static fromField(x: Field, y: Field) {
+    return new Point(x, y);
+  }
+
+  static addPoints(a: Point, b: Point) {
+    return new Point(a.x.add(b.x), a.y.add(b.y));
+  }
+}
+
+const point1 = Point.fromField(Field(10), Field(4));
+const point2 = Point.fromField(Field(1), Field(2));
+
+const pointSum = Point.addPoints(point1, point2);
+
+console.log(`pointSum Fields: ${pointSum.toFields().map((p) => p.toString())}`);
+
+class Points8 extends CircuitValue {
+  @arrayProp(Point, 8) points: Point[];
+
+  constructor(points: Point[]) {
+    super();
+    this.points = points;
+  }
+}
+
+const pointsArray = new Array(8).fill(null).map((_, i) => Point.fromField(Field(i), Field(i*10)));
+const points8 = new Points8(pointsArray);
+
+console.log(`points8 Fields: ${JSON.stringify(points8)}`);
+```
+
+Which prints the following to the console:
+
+```sh
+pointSum Fields: 11,6
+points8 Fields: {"points":[{"x":"0","y":"0"},{"x":"1","y":"10"},{"x":"2","y":"20"},{"x":"3","y":"30"},{"x":"4","y":"40"},{"x":"5","y":"50"},{"x":"6","y":"60"},{"x":"7","y":"70"}]}
+```
+
+A great example of a complex use of CircuitValue is the MerkleTree implementation  - see the source for that [here](https://github.com/o1-labs/snarkyjs/blob/main/src/lib/merkle_tree.ts) for more.
+
+## Control flow
+
+There are two functions which help do control flow within SnarkyJS:
+
+* [Circuit.if](../snarkyjs-reference/classes/Circuit#if)
+* [Circuit.switch](../snarkyjs-reference/classes/Circuit#switch)
+
+With these, you can write conditionals inside SnarkyJS.
+
+For example:
+
+```ts
+const input1 = Int64.from(10);
+const input2 = Int64.from(-15);
+
+const inputSum = input1.add(input2);
+
+const inputSumAbs = Circuit.if(inputSum.isPositive(), inputSum, inputSum.mul(Int64.minusOne));
+
+console.log(`inputSum: ${inputSum.toString()}`);
+console.log(`inputSumAbs: ${inputSumAbs.toString()}`);
+
+const input3 = Int64.from(22);
+
+const input1largest = input1.sub(input2).isPositive().and(input1.sub(input3).isPositive());
+const input2largest = input2.sub(input1).isPositive().and(input2.sub(input3).isPositive());
+const input3largest = input3.sub(input1).isPositive().and(input3.sub(input2).isPositive());
+
+const largest = Circuit.switch([ input1largest, input2largest, input3largest ], Int64, [ input1, input2, input3 ])
+
+console.log(`largest: ${largest.toString()}`);
+```
+
+With output:
+
+```sh
+inputSum: -5
+inputSumAbs: 5
+largest: 22
+```
+
+By inserting function calls, you can write circuits which usual if / else statements:
+
+```ts
+const result = Circuit.if(
+  <condition>, 
```

Comment:
This might be easier to read if it avoided non-valid syntax and relied on a basic example (like a equals b) and comments for things like `// condition`

Otherwise I stopped to think, is this TS syntax, and then realized I just needed to mentally parse it.